carstore/bs.go

```diff
@@ -1411,35 +1449,42 @@ func (cs *CarStore) deleteStaleRefs(ctx context.Context, brefs []blockRef, stale
 		brByCid[br.Cid.CID] = append(brByCid[br.Cid.CID], br)
 	}
 
-	var staleToDelete []uint
+	var staleToKeep []cid.Cid
 	for _, sr := range staleRefs {
-		brs := brByCid[sr.Cid.CID]
-		del := true
-		for _, br := range brs {
-			if !removedShards[br.Shard] {
-				del = false
-				break
-			}
+		cids, err := sr.getCids()
+		if err != nil {
+			return fmt.Errorf("getCids on staleRef failed (%d): %w", sr.ID, err)
 		}
 
-		if del {
-			staleToDelete = append(staleToDelete, sr.ID)
+		for _, c := range cids {
+			brs := brByCid[c]
+			del := true
+			for _, br := range brs {
+				if !removedShards[br.Shard] {
+					del = false
+					break
+				}
+			}
+
+			if !del {
+				staleToKeep = append(staleToKeep, c)
+			}
 		}
 	}
 
-	chunkSize := 10000
-	for i := 0; i < len(staleToDelete); i += chunkSize {
-		sl := staleToDelete[i:]
-		if len(sl) > chunkSize {
-			sl = sl[:chunkSize]
-		}
+	if err := cs.meta.Delete(&staleRef{}, "usr = ?", uid).Error; err != nil {
+		return err
+	}
```

Comment:
Do we want to wrap this in a transaction with the new staleref creation? Is there a potential race condition here or do we only create `staleRefs` in this function? I can't remember if we also create them when a record gets deleted if any relevant blocks have no more refs.